### PR TITLE
Introducing new `llvm_pattern_matching_failure_event` to Proof Events

### PR DESCRIPTION
**Pull request by Robertorosmaninho**
Fixes: https://github.com/Pi-Squared-Inc/pi2/issues/1722
This PR introduces a new function event called `llvm_pattern_matching_failure_event`, which is identified by the `pattern_matching_failure_sentinel`. This new event flags to the user that an attempt to apply a rule failed in pattern matching its constructor; therefore, no further simplification can be done at this point.

### CODE REVIEW

**Comment by Robertorosmaninho** [comment]
We still need to bump the version from 11 to 12, but this will be terrible to review after modifying all the tests, so I'll bump it as soon as I have your approval in the main implementation.

**Review by dwightguth — APPROVED** [review]

File: docs/proof-trace.md
Comment (theo25) [inline docs/proof-trace.md:53]: This needs to appear in the right hand side of some production, most likely the `argument` and `event` productions.

**Review by theo25 — CHANGES_REQUESTED** [review]
Looks good, but please take care of the one comment that I left.

**Review by theo25 — APPROVED** [review]

**Pull request merged**
